src/main/java/org/opensearch/knn/index/codec/KNN990Codec/KNNQuantizationStateReader.java

```diff
@@ -0,0 +1,143 @@
+import org.apache.lucene.index.SegmentReadState;
+import org.apache.lucene.store.Directory;
+import org.apache.lucene.store.IOContext;
+import org.apache.lucene.store.IndexInput;
+import org.opensearch.knn.common.KNNConstants;
+import org.opensearch.knn.quantization.models.quantizationState.QuantizationState;
+
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+
+/**
+ * Reads quantization states
+ */
+public class KNNQuantizationStateReader {
+
+    /**
+     * Read quantization states and return list of fieldNames and bytes
+     * File format:
+     * Header
+     * QS1 state bytes
+     * QS2 state bytes
+     * Number of quantization states
+     * QS1 field name
+     * QS1 state bytes length
+     * QS1 position of state bytes
+     * QS2 field name
+     * QS2 state bytes length
+     * QS2 position of state bytes
+     * Position of index section (where QS1 field name is located)
+     * -1 (marker)
+     * Footer
+     *
+     * @param state the read state to read from
+     */
+    public Map<String, byte[]> read(SegmentReadState state) throws IOException {
+        String quantizationStateFileName = IndexFileNames.segmentFileName(
+            state.segmentInfo.name,
+            state.segmentSuffix,
+            KNNConstants.QUANTIZATION_STATE_FILE_SUFFIX
+        );
+        Map<String, byte[]> readQuantizationStateInfos = new HashMap<>();
+
+        IndexInput input = state.directory.openInput(quantizationStateFileName, IOContext.READ);
+        CodecUtil.retrieveChecksum(input);
+
+        int numFields = getNumFields(input);
+
+        List<String> fieldNames = new ArrayList<>();
+        List<Long> positions = new ArrayList<>();
+        List<Integer> lengths = new ArrayList<>();
+
+        // Read each field's metadata from the index section
+        for (int i = 0; i < numFields; i++) {
+            fieldNames.add(input.readString());
+            int length = input.readInt();
+            lengths.add(length);
+            long position = input.readVLong();
+            positions.add(position);
+        }
+        // Read each field's bytes
+        for (int i = 0; i < numFields; i++) {
+            input.seek(positions.get(i));
+            byte[] stateBytes = new byte[lengths.get(i)];
+            input.readBytes(stateBytes, 0, lengths.get(i));
+            readQuantizationStateInfos.put(fieldNames.get(i), stateBytes);
+        }
+        input.close();
+        return readQuantizationStateInfos;
+    }
+
+    /**
+     * Reads an individual quantization state for a given field
+     * @param directory directory to open input
+     * @param segmentName segment name
+     * @param segmentSuffix segment suffix
+     * @param fieldInfo field information
+     * @return quantization state
+     */
+    public QuantizationState read(Directory directory, String segmentName, String segmentSuffix, FieldInfo fieldInfo) throws IOException {
```

Comment:
Let's wrap all four arguments in Model , and Opening of Index Input Let's have in Constructor  for both reads.